aten/src/ATen/ParallelNative.cpp

```diff
@@ -125,13 +125,12 @@ namespace internal {
 
 inline std::tuple<size_t, size_t> calc_num_tasks_and_chunk_size(
     int64_t begin, int64_t end, int64_t grain_size) {
-  if ((end - begin) < grain_size) {
-    return std::make_tuple(1, std::max((int64_t)0, end - begin));
-  }
+  int64_t num_threads = get_num_threads();
   // Choose number of tasks based on grain size and number of threads.
-  size_t chunk_size = divup((end - begin), get_num_threads());
-  // Make sure each task is at least grain_size size.
```

Comment:
Can you explain how this breaks your code? I can't see how a different `chunk_size` could give you the wrong answer unless your code is making bad assumptions to begin with.

Reply:
I make assumptions based on OMP strategy, that is I create buffers per thread.

Reply:
If your code breaks with anything other than one exact partitioning strategy, then IMO you should do that partitioning in your code, and only use `parallel_for` to iterate over chunks.

Reply:
That will work in relation to my problem, but what is the benefit of having different partition strategies for different backends? I see no point in being divergent when IMO they could implement the same approach.

Reply:
Btw, if you compare the strategies closer, you will not be able to implement a uniform strategy, just because the number of threads and chunk size is decided inside the parallel_for loop.

Reply:
I don't think it makes sense to intentionally write diverging strategies, but I also don't think it's robust for code to assume internal implementation details of `parallel_for` that may or may not hold in general. For example, the underlying OpenMP implementation may choose to use fewer than `omp_get_max_threads()` in the parallel region, and suddenly your code breaks.

Reply:
If OpenMP does not provide any guaranties on the topology, how one would now about the pool of threads available?

Reply:
`at::get_num_threads` in OpenMP is equivalent to `omp_get_max_threads()` so as the name suggests, it is indeed the maximum number of threads that may be used.

>this number seems to guide how OMP backend is handling parallelism.

The OpenMP backend calculates the chunk size within the parallel region, where `omp_get_num_threads()` returns the exact number of threads used for that parallel region. Specifically to avoid the aforementioned issues.

Reply:
Right, so could it happen that the same work processed several times will receive different number of threads per each processing step?

Reply:
I'm not 100% sure to be honest. I don't know if that's explicitly disallowed by the OpenMP standard, but I expect that the number of threads would be chosen deterministically which should mean the same both times.

Reply:
Makes sense. I will take your advice and reimplement the algorithms in a non-backend-dependent way. Thank you for this insightful discussion!